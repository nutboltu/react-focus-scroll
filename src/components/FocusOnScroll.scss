.focusOnScroll {
  display: block;

  .section {
    opacity: 0.5;
    transition: opacity 0.1s ease-in-out;
  }

  .isFocused {
    opacity: 1;
  }
}