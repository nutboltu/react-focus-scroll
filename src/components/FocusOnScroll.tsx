import * as React from 'react';
import { throttle, debounce } from 'lodash';

import {
  DELAY_TIME_IN_MS,
  SUPPORTED_TAGS_FOR_FOCUS,
} from '../helpers/config';
import {
  findFocusArea,
  findFocusElement,
} from '../helpers/utils';

import {
  IFocusOnScrollProps,
  IFocusOnScrollState,
  ChildrenArray,
} from '../@types';

class FocusOnScroll extends React.Component<IFocusOnScrollProps, IFocusOnScrollState> {
  private scrollElement: HTMLElement;

  constructor(props: IFocusOnScrollProps) {
    super(props);
    this.state = {
      focusIndex: Number(this.props.focusOn) || 0,
      focusArea: {
        top: 0,
        bottom: 0,
      },
      focusOnTriggered: false,
      childrenArray: [],
    };
  }
  private setScrollElement = (element: HTMLDivElement): void => {
    this.scrollElement = element;
  }
  private debounceResizeHandler = () => debounce(this.resizeHandler, DELAY_TIME_IN_MS);
  private throttleScrollHandler = () => throttle(this.scrollHandler, DELAY_TIME_IN_MS);

  private findAndSetChildren = (callback: unknown): void => {
    if (this.scrollElement && this.scrollElement.children) {
      const childrenArray = Array.from(this.scrollElement.children) as ChildrenArray;
      this.setState({ childrenArray }, () => {
        if (typeof callback === 'function') {
          callback();
        }
      });
    }
  }

  private getInlineStyles = (focused: boolean) => {
    return {
      opacity: focused ? 1 : 0.5,
      transition: 'opacity 0.1s ease-in-out',
    };
  }

  public componentDidMount() {
    this.findAndSetChildren(this.resizeHandler);
    window.addEventListener('scroll', this.throttleScrollHandler());
    window.addEventListener('resize', this.debounceResizeHandler());
  }

  public componentWillReceiveProps(nextProps: IFocusOnScrollProps) {
    if (this.props.focusOn !== nextProps.focusOn) {
      this.focusOnHandler(nextProps.focusOn!);
    }
  }

  public componentWillUnmount() {
    window.removeEventListener('scroll', this.throttleScrollHandler);
    window.removeEventListener('resize', this.debounceResizeHandler);
  }

  public focusOnHandler = (index: string): void => {
    const { focusArea, childrenArray } = this.state;
    const focusIndex = Number(index);
    if (isNaN(focusIndex) || focusIndex > childrenArray.length) {
      return ;
    }

    const element = childrenArray[focusIndex] as HTMLElement;
    const scrollPosition = (element.getBoundingClientRect().top + window.pageYOffset) - focusArea.top;
    this.setState({
      focusOnTriggered: true,
    }, () => {
      window.scrollTo(0, scrollPosition + 1);
      setTimeout(() => {
        this.setState({ focusOnTriggered: false });
      }, DELAY_TIME_IN_MS);
      this.setFocusIndex(focusIndex);
    });
  }

  /**
   * If user clicks on an input field of a blur section
   *  this handler will automatically focus that section
   */
  public inputClickHandler = (event: any): void => {
    const tagName = event.target && event.target.tagName;
    const unsuportedTag = (!tagName || !SUPPORTED_TAGS_FOR_FOCUS.includes(tagName.toLowerCase()));

    if (unsuportedTag) {
      return;
    }

    const { childrenArray, focusIndex } = this.state;
    const childElement = event.target;
    const targetIndex = childrenArray.findIndex((parentElement: HTMLElement) => parentElement.contains(childElement));
    const alreadyInFocusArea = focusIndex === targetIndex;
    if (!alreadyInFocusArea) {
      childElement.scrollIntoView({ behavior: 'smooth' });
      this.setFocusIndex(targetIndex || focusIndex);
    }
  }

  public resizeHandler = (): void => {
    const { childrenArray } = this.state;
    const focusArea = findFocusArea(childrenArray);
    this.setState({
      focusArea,
    }, this.scrollHandler);
  }

  public scrollHandler = (): void => {
    const { focusArea, childrenArray, focusOnTriggered } = this.state;
    if (focusOnTriggered) {
      return;
    }
    const focusIndex = findFocusElement(focusArea, childrenArray);
    if (focusIndex >= 0) {
      this.setFocusIndex(focusIndex);
    }
  }

  public setFocusIndex = (focusIndex: number): void => {
    if (this.state.focusIndex !== focusIndex) {
      const { onFocus } = this.props;
      this.setState({
        focusIndex,
      });
      if (typeof onFocus === 'function') {
        onFocus(focusIndex);
      }
    }
  }

  public renderSections = (): React.ReactNode => {
    const {
      children,
    } = this.props;

    return React.Children.map(children, (element, index) => {
      const focused = this.state.focusIndex === index;
      const focusedClassName = focused ? 'focused' : '';
      const id = `rfs-section-${index}`;

      return (
        <section
          key={id}
          style={this.getInlineStyles(focused)}
          className={focusedClassName}
        >
          {element}
        </section>
      );
    });
  }

  public render(): React.ReactNode {
    return (
      <div
        className={this.props.className || ''}
        role='button'
        ref={this.setScrollElement}
        onClick={this.inputClickHandler}
      >
        {this.renderSections()}
      </div>
    );
  }
}

export default FocusOnScroll;
