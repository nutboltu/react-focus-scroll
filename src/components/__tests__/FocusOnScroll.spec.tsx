import * as React from 'react';

import { shallow } from 'enzyme';
import { spy } from 'sinon';
import FocusOnScroll from '../FocusOnScroll';

describe('FocusOnScroll Component Suite', () => {
  const getComponent = () => shallow(<FocusOnScroll >
    <div> This is a test</div>
    <div> This is a test</div>
  </FocusOnScroll>);

  describe('FocusOnScroll Component', () => {
    test('should render with defaults', () => {
      expect(getComponent()).toMatchSnapshot();
    });

    it('should render 2 sections', () => {
      const sections = getComponent().find('section');
      expect(sections).toHaveLength(2);
    });
  });

  describe('inputClickHandler', () => {
    it('should call scrollIntoView for clicked element', () => {
      const component = getComponent();
      component.setState({
        focusIndex: 0,
        childrenArray: [{
          offsetTop: 0,
          clientHeight: 100,
          contains: () => false,
        }, {
          offsetTop: 101,
          clientHeight: 100,
          contains: () => true,
        }],
      });
      const instance = component.instance() as any;
      const mockScrollIntoView = spy();
      const event = {
        target: {
          getBoundingClientRect: () => ({ top: 102 }),
          tagName: 'INPUT',
          scrollIntoView: mockScrollIntoView,
        },
      };

      instance.inputClickHandler(event);
      expect(mockScrollIntoView.calledOnce).toBeTruthy;
    });
    it('shouldn\'t call scrollIntoView if clicks on DIV element', () => {
      const component = getComponent();
      const mockScrollIntoView = spy();
      component.setState({
        focusIndex: 0,
        childrenArray: [{
          offsetTop: 0,
          clientHeight: 100,
        }, {
          offsetTop: 101,
          clientHeight: 100,
        }],
      });
      const instance = component.instance() as any;
      const event = {
        target: {
          getBoundingClientRect: () => ({ top: 102 }),
          tagName: 'DIV',
          scrollIntoView: mockScrollIntoView,
        },
      };

      instance.inputClickHandler(event);
      expect(mockScrollIntoView.calledOnce).toBeFalsy;
    });
    it('shouldn\'t call scrollIntoView if child is already in focused area', () => {
      const component = getComponent();
      component.setState({
        focusIndex: 1,
        childrenArray: [{
          offsetTop: 0,
          clientHeight: 100,
          contains: () => false,
        }, {
          offsetTop: 101,
          clientHeight: 100,
          contains: () => true,
        }],
      });
      const instance = component.instance() as any;
      const mockScrollIntoView = spy();
      const event = {
        target: {
          getBoundingClientRect: () => ({ top: 10 }),
          tagName: 'DIV',
          scrollIntoView: mockScrollIntoView,
        },
      };

      instance.inputClickHandler(event);
      expect(mockScrollIntoView.calledOnce).toBeFalsy;
    });
  });
});
