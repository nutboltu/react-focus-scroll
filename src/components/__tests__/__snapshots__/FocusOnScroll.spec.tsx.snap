// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FocusOnScroll Component Suite FocusOnScroll Component should render with defaults 1`] = `
<div
  className=""
  onClick={[Function]}
  role="button"
>
  <section
    className="focused"
    key="rfs-section-0/.0"
    style={
      Object {
        "opacity": 1,
        "transition": "opacity 0.1s ease-in-out",
      }
    }
  >
    <div>
       This is a test
    </div>
  </section>
  <section
    className=""
    key="rfs-section-1/.1"
    style={
      Object {
        "opacity": 0.5,
        "transition": "opacity 0.1s ease-in-out",
      }
    }
  >
    <div>
       This is a test
    </div>
  </section>
</div>
`;
